airbyte-integrations/connectors/source-posthog/source_posthog/streams.py

```diff
@@ -0,0 +1,307 @@
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def get_updated_state(
+        self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]
+    ) -> Union[Mapping[str, Any], Tuple[Any, Mapping[str, Any]]]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        current_stream_state = current_stream_state or {}
+        current_state = current_stream_state.get(self.cursor_field)
+        latest_state = latest_record.get(self.cursor_field)  # 810->809->808
+
+        if current_state and latest_state <= current_state:
+            # return Tuple with any first element and current state
+            return (-1, self.reversed_pagination["upgrade_to"])
+        if (
+            self.reversed_pagination["latest_response_state"]
+            and latest_state <= self.reversed_pagination["latest_response_state"][self.cursor_field]
+            and self.reversed_pagination["is_completed"]
+        ):
+            return self.reversed_pagination["upgrade_to"]
+
+        return current_stream_state  # dont upgrade state until complete!
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        data = response_json[self.data_field]
+        if data:
+            if not ("next" in response_json and response_json["next"]):
+                self.reversed_pagination["is_completed"] = True
+            latest_field_value = data[-1][self.cursor_field]
+            if stream_state:
+                if latest_field_value <= stream_state.get(self.cursor_field):
+                    self.reversed_pagination["upgrade_to"] = stream_state
+                    yield from []
+                else:
+                    self.reversed_pagination["latest_response_state"] = stream_state
+
+            else:
+                self.reversed_pagination["latest_response_state"] = {self.cursor_field: latest_field_value}
+            if not self.reversed_pagination["upgrade_to"]:
+                self.reversed_pagination["upgrade_to"] = {self.cursor_field: data[0][self.cursor_field]}
+
+        yield from data
+
+
+class Annotations(PosthogStream):
+    """
+    Could not use inremental here. No option to sort in request provided by API.
+    There are 3 possible key fields id, created_at, updated_at.
+    The last one is the most suitable
+    But response is always sorted DESC on `id or created_at`.
+    If the field was updated and `updated_at` was increased, it did not change its position in response.
+    So if we have 2766-2891 ids, state saved on id=2870 and update 2869th record,
+    we'll miss the update. So no logical way to use incremental, it will be like full_refresh.
+    """
```

Comment:
But, if we can sort by updated_at, so we can store all states of the record as before its change, and after the change.

Reply:
we cannot sort by updated_at during request. In result we have unordered answer, also paginated. I will need to query ALL the data, sort all the responces and it will be factually full-refresh

Reply:
But i see in docs, that sort by `updated_at `is real: https://posthog.com/docs/api/annotations#endpoints
No?